runners/samza/src/main/java/org/apache/beam/runners/samza/renderer/PipelineJsonRenderer.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.renderer;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.samza.SamzaIOInfo;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.runners.TransformHierarchy;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JSON renderer for BEAM {@link Pipeline} DAG. This can help us with visualization of the Beam
+ * DAG.
+ */
+@Experimental
+public class PipelineJsonRenderer implements Pipeline.PipelineVisitor {
+  private static final Logger LOG = LoggerFactory.getLogger(PipelineJsonRenderer.class);
+  private static final String OUTERMOST_NODE = "OuterMostNode";
+
+  public static String toJsonString(Pipeline pipeline) {
+    final PipelineJsonRenderer visitor = new PipelineJsonRenderer();
+    visitor.begin();
+    pipeline.traverseTopologically(visitor);
+    visitor.end();
+    return visitor.jsonBuilder.toString();
+  }
+
+  public static Optional<Object> toJsonString(RunnerApi.Pipeline pipeline) {
+    return Optional.empty();
+  }
+
+  private final StringBuilder jsonBuilder = new StringBuilder();
+  private final StringBuilder graphLinks = new StringBuilder();
+  private final Map<TransformHierarchy.Node, Integer> nodeToId = new HashMap<>();
+  private final Map<PValue, String> valueToProducerNodeName = new HashMap<>();
+  private int indent;
+  private int nextNodeId;
+
+  private PipelineJsonRenderer() {}
+
+  @Override
+  public void enterPipeline(Pipeline p) {}
+
+  @Override
+  public CompositeBehavior enterCompositeTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine(
+        "{ \"fullName\":\"%s\",", escapeString(fullName.isEmpty() ? OUTERMOST_NODE : fullName));
+    if (!fullName.isEmpty()) {
```

Comment:
check on `node.getEnclosingNode()` is more readable.  fullName.isEmpty() is based on the assumption that only outer most node has empty full name. It is probably functionally the same but `node.getEnclosingNode() != null` is easier for others to understand.

Reply:
Sure, made the change.